### PR TITLE
adding ClusTree clustering extension

### PR DESCRIPTION
**Pull request by niruhan**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -25,3 +25,6 @@ target
 # Ignore target directories
 target/
 *.*~
+
+org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/a1.csv
+TryMOA.class
```
Comment (nirmal070125) [inline .gitignore]: Why are these added?

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+                                "which the current event belongs to.",
+                        type = {DataType.DOUBLE}
+                )
+        },
+        examples = {
+                @Example(
+                        syntax = "@App:name('ClusTreeTestSiddhiApp') \n" +
+                                "define stream InputStream (x double, y double);\n" +
+                                "@info(name = 'query1') \n" +
+                                "from InputStream#streamingml:ClusTree(2, 10, 20, 5, 50, x, y) \n" +
```
Comment (nirmal070125) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java]: should be lower case 'c' right?

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+                                "select closestCentroidCoordinate1, closestCentroidCoordinate2, x, y \n" +
+                                "insert into OutputStream;",
+                        description = "This query will create a Siddhi app named ClusTreeTestSiddhiApp and will " +
+                                "accept 2D inputs od doubles. The query which is named query1 will create a ClusTree " +
+                                "model and will create a kmeans model after firsat 20 events and will " +
+                                "refresh it every 20 events after. Number of macro clusters will be 2 and the " +
+                                "maximum iterations of kmeans to converge will be 10. The max height of tree is " +
+                                "set to 5 so at maximum we will get 3^5 micro clusters from ClusTree and the " +
+                                "horizon is set as 50, so after 50 events micro clusters that were not updated " +
+                                "will lose their weight by half." //todo: add eg with default params
```
Comment (nirmal070125) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java]: todo?

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+    private ExecutorService executorService;
+    private ClusTreeModel clusTreeModel;
+    private KMeansModel kMeansModel;
+    private List<VariableExpressionExecutor> featureVariableExpressionExecutors = new LinkedList<>();
+    private static final Logger logger = Logger.getLogger(ClusTreeStreamProcessorExtension.class.getName());
+
+    @Override
+    protected List<Attribute> init(AbstractDefinition abstractDefinition, ExpressionExecutor[] expressionExecutors,
+                                   ConfigReader configReader, SiddhiAppContext siddhiAppContext) {
+        int noOfFeatures = inputDefinition.getAttributeList().size();
```
Comment (nirmal070125) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java]: this is maxNoOfFeatures, not # of features.

**Review by nirmal070125 — CHANGES_REQUESTED** [review]
added some comments.

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+        examples = {
+                @Example(
+                        syntax = "@App:name('ClusTreeTestSiddhiApp') \n" +
+                                "define stream InputStream (x double, y double);\n" +
+                                "@info(name = 'query1') \n" +
+                                "from InputStream#streamingml:ClusTree(2, 10, 20, 5, 50, x, y) \n" +
+                                "select closestCentroidCoordinate1, closestCentroidCoordinate2, x, y \n" +
+                                "insert into OutputStream;",
+                        description = "This query will create a Siddhi app named ClusTreeTestSiddhiApp and will " +
+                                "accept 2D inputs od doubles. The query which is named query1 will create a ClusTree " +
```
Comment (tishan89) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java]: typo 'od doubles'

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+                                "querying is suggested for users who are not familier with ClusTree/KMeans algorithms."
+                )
+        }
+)
+public class ClusTreeStreamProcessorExtension extends StreamProcessor {
+    private final int minConstantParams = 1;
+    private final int maxConstantParams = 5;
+    private final int separateThreadThreshold = 500;
+    private int noOfClusters;
+    private int noOfEventsToRefreshMacroModel = 100; //todo : tune. check algo defaults
```
Comment (tishan89) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java]: resolve todo

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+                double minHeightOfTree = (Math.log(noOfClusters) / Math.log(3));
+                if (maxHeightOfTree < minHeightOfTree) {
+                    throw new SiddhiAppCreationException("maxHeightOfTree should be an int greater than " +
+                            minHeightOfTree + " but found " + maxHeightOfTree);
+                }
+            } else {
+                throw new SiddhiAppCreationException("maxHeightOfTree should be of type int but found " +
+                        attributeExpressionExecutors[3].getReturnType());
+            }
+            maxHeightOfTree -= 1; // since this parameter when set to 0 MOA ClusTree implementation builds one root
```
Comment (tishan89) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java]: comment is not clear enough

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+                           StreamEventCloner streamEventCloner, ComplexEventPopulater complexEventPopulater) {
+        synchronized (this) {
+            while (complexEventChunk.hasNext()) {
+                StreamEvent streamEvent = complexEventChunk.next();
+                noOfEventsReceived++;
+
+                //validating and getting coordinate values
+                for (int i = 0; i < noOfDimensions; i++) {
+                    try {
+                        Number content = (Number) featureVariableExpressionExecutors.get(i).execute(streamEvent);
```
Comment (tishan89) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java:321]: Can't we do this validation at init() by using getReturnType() method in ExpresionExecutor?

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java [hunk 9]
```diff
@@ -0,0 +1,388 @@
...
+                    } else {
+                        Trainer trainer = new Trainer(kMeansModel, dpa, noOfClusters, maxIterations, noOfDimensions);
+                        Future f = executorService.submit(trainer);
+                    }
+                }
+
+                //make prediction if the model is trained
+                if (kMeansModel.isTrained()) {
+                    if (logger.isDebugEnabled()) {
+                        logger.debug("Populating output");
```
Comment (tishan89) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java]: can we add output?

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/util/ClusTreeModel.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+    private synchronized Instance createMOAInstance(double[] cepEvent) {
+        Instance instance = new DenseInstance(1.0D, cepEvent);
+        return instance;
+    }
+
+    public synchronized Clustering getMicroClustering() {
+        return clusTree.getMicroClusteringResult();
+    }
+
+    public synchronized LinkedList<DataPoint> getMicroClusteringAsDPArray() {
```
Comment (tishan89) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/util/ClusTreeModel.java]: Why do you need to return LinkedList instead of a List

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/util/ClusTreeModel.java [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+    /**
+     * @param cepEvent Event Data
+     * @return represents a single Event
+     */
+    private synchronized Instance createMOAInstance(double[] cepEvent) {
+        Instance instance = new DenseInstance(1.0D, cepEvent);
+        return instance;
+    }
+
+    public synchronized Clustering getMicroClustering() {
```
Comment (tishan89) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/util/ClusTreeModel.java]: And why do we need all these synchonization?

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/util/Trainer.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+ */
+public class Trainer implements Runnable {
+    private KMeansModel kMeansModel;
+    private LinkedList<DataPoint> dataPointsArray;
+    private int noOfClusters;
+    private int maxIterations;
+    private int noOfDimensions;
+
+
+    public Trainer(KMeansModel kMeansModel, LinkedList<DataPoint> dataPointsArray, int noOfClusters, int maxIterations,
```
Comment (tishan89) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/util/Trainer.java]: linkedList -> List

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/util/WeightedKMeans.java [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * perform weighted kmeans calculations. a singleton class
+ */
+public class WeightedKMeans {
+    private static final Logger logger = Logger.getLogger(WeightedKMeans.class.getName());
+
+    public static List<Cluster> run(LinkedList<DataPoint> dataPointsArray, int noOfClusters, int maximumIterations,
```
Comment (tishan89) [inline component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/util/WeightedKMeans.java]: LinkedList -> List

File: component/src/test/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtensionTest.java [hunk 14]
```diff
@@ -0,0 +1,406 @@
...
+                        "from InputStream#streamingml:clusTree(2, 10, 20, 5, 50, x, y) " +
+                        "select closestCentroidCoordinate1, closestCentroidCoordinate2, x, y " +
+                        "insert into OutputStream;");
+        SiddhiAppRuntime siddhiAppRuntime = siddhiManager.createSiddhiAppRuntime(inputStream + query);
+
+        siddhiAppRuntime.addCallback("query1", new QueryCallback() {
+            @Override
+            public void receive(long timeStamp, Event[] inEvents, Event[] removeEvents) {
+                EventPrinter.print(inEvents);
+                /*for (Event event: inEvents) {
```
Comment (tishan89) [inline component/src/test/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtensionTest.java]: can't you assert the output?

**Review by tishan89 — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (niruhan) [reply .gitignore]: i added some classes to test moa lib then removed. ill delete

Comment (niruhan) [reply .gitignore]: done

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java (hunk 2 above)
Comment (niruhan) [reply component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java]: fixed

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java (hunk 5 above)
Comment (niruhan) [reply component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java]: fixed

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java (hunk 6 above)
Comment (niruhan) [reply component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java]: fixed

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java (hunk 7 above)
Comment (niruhan) [reply component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java]: fixed

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java (hunk 8 above)
Comment (niruhan) [reply component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java:321]: we are doing the validation of the query feature type in init. but this is done to check if the user fails to honor the stream definition and pass a non-numeric into the inputhandler

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/util/ClusTreeModel.java (hunk 11 above)
Comment (niruhan) [reply component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/util/ClusTreeModel.java]: we dont need this. will remove

Comment (niruhan) [reply component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/util/ClusTreeModel.java]: fixed

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java (hunk 9 above)
Comment (niruhan) [reply component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java]: sorry i dont understand. please explain

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/util/ClusTreeModel.java (hunk 10 above)
Comment (niruhan) [reply component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/util/ClusTreeModel.java]: fixed

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/util/Trainer.java (hunk 12 above)
Comment (niruhan) [reply component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/util/Trainer.java]: fixed

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/util/WeightedKMeans.java (hunk 13 above)
Comment (niruhan) [reply component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/util/WeightedKMeans.java]: fixed

File: component/src/test/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtensionTest.java (hunk 14 above)
Comment (niruhan) [reply component/src/test/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtensionTest.java]: done

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java (hunk 3 above)
Comment (niruhan) [reply component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java]: fixed

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java (hunk 4 above)
Comment (niruhan) [reply component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java]: fixed

File: component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java (hunk 9 above)
Comment (nirmal070125) [reply component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java]: basically explain what is output. Always think from the person who sees this log. What can he get to know from this?

Comment (niruhan) [reply component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java]: oh this will populate the event with the prediction once the model is trained. i will add the details into the log

**Review by nirmal070125 — APPROVED** [review]

**Review by tishan89 — APPROVED** [review]
